<a href="https://colab.research.google.com/github/nguemo12/AI-Beckn-Recommendation-system/blob/master/pw4_eval_ie.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a target="_blank" href="https://colab.research.google.com/github/PaulLerner/aivancity_nlp/blob/main/pw4_eval_ie.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

# Installation and imports

Hit `Ctrl+S` to save a copy of the Colab notebook to your drive

Run on Google Colab GPU:
- Connect
- Modify execution
- GPU

![image.png](https://paullerner.github.io/aivancity_nlp/_static/colab_gpu.png)

In [ ]:
import torch

from torch import nn

from transformers import AutoTokenizer, AutoModel

/home/paul/miniforge3/envs/matos/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [ ]:
assert torch.cuda.is_available(), "Connect to GPU and try again (ask teacher for help)"

# Data

We'll use the [E3C dataset](https://books.openedition.org/aaccademia/pdf/8663) of Named Entity Recognition in clinical texts.

Therefore, we'll use a model pretrained on scientific texts: [SciBert](https://www.aclweb.org/anthology/D19-1371)

## Loading


In [ ]:
!wget https://raw.githubusercontent.com/hltfbk/E3C-Corpus/refs/heads/main/preprocessed_data/clinical_entities/layer1/English/test.txt

--2024-11-21 16:21:46--  https://raw.githubusercontent.com/hltfbk/E3C-Corpus/refs/heads/main/preprocessed_data/clinical_entities/layer1/English/test.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.108.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


200 OK
Length: 136768 (134K) [text/plain]
Saving to: 'test.txt'

test.txt            100%[===================>] 133.56K  --.-KB/s    in 0.03s   

2024-11-21 16:21:47 (4.57 MB/s) - 'test.txt' saved [136768/136768]



In [ ]:
!wget https://raw.githubusercontent.com/hltfbk/E3C-Corpus/refs/heads/main/preprocessed_data/clinical_entities/layer1/English/train.txt

--2024-11-21 16:21:47--  https://raw.githubusercontent.com/hltfbk/E3C-Corpus/refs/heads/main/preprocessed_data/clinical_entities/layer1/English/train.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.108.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


200 OK
Length: 103512 (101K) [text/plain]
Saving to: 'train.txt'

train.txt           100%[===================>] 101.09K  --.-KB/s    in 0.01s   

2024-11-21 16:21:47 (6.99 MB/s) - 'train.txt' saved [103512/103512]



In [ ]:
encoder = AutoModel.from_pretrained("allenai/scibert_scivocab_cased", device_map="auto", add_pooling_layer=False)

In [ ]:
encoder

BertModel(
  (embeddings): BertEmbeddings(
    (word_embeddings): Embedding(31116, 768, padding_idx=0)
    (position_embeddings): Embedding(512, 768)
    (token_type_embeddings): Embedding(2, 768)
    (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (encoder): BertEncoder(
    (layer): ModuleList(
      (0-11): 12 x BertLayer(
        (attention): BertAttention(
          (self): BertSdpaSelfAttention(
            (query): Linear(in_features=768, out_features=768, bias=True)
            (key): Linear(in_features=768, out_features=768, bias=True)
            (value): Linear(in_features=768, out_features=768, bias=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (output): BertSelfOutput(
            (dense): Linear(in_features=768, out_features=768, bias=True)
            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
            (dropout): Dropout(p=0.1, inplace=False

In [ ]:
tokenizer = AutoTokenizer.from_pretrained("allenai/scibert_scivocab_cased", do_lower_case=False)

In [ ]:
LABELS = {'O': 0, 'B-ety': 1, 'I-ety': 2}
i2label = {i: label for label, i in LABELS.items()}

In [ ]:
def read_data(path):
    with open(path,"rt") as file:
        lines = file.read().strip().split("\n")
    dataset = []
    tokens, labels = [], []
    for line in lines:
        if line:
            token, label = line.split(" ")
            tokens.append(token)
            labels.append(LABELS[label])
        else:
            dataset.append((tokens, labels))
            tokens, labels = [], []
    dataset.append((tokens, labels))
    return dataset
train_set = read_data("train.txt")

In [ ]:
test_set = read_data("test.txt")

In [ ]:
print(len(train_set), len(test_set))

669 851


In [ ]:
item = train_set[0]
print(" ".join(item[0]))
print(list(zip(item[0], [i2label[label] for label in item[1]])))

A 14 - year old boy with no significant past medical history presented to a small district hospital in southern Sierra Leone with a 4 day history of facial puffiness , peripheral pitting oedema , abdominal pains , and reduced urine output .
[('A', 'O'), ('14', 'O'), ('-', 'O'), ('year', 'O'), ('old', 'O'), ('boy', 'O'), ('with', 'O'), ('no', 'O'), ('significant', 'O'), ('past', 'O'), ('medical', 'O'), ('history', 'O'), ('presented', 'O'), ('to', 'O'), ('a', 'O'), ('small', 'O'), ('district', 'O'), ('hospital', 'O'), ('in', 'O'), ('southern', 'O'), ('Sierra', 'O'), ('Leone', 'O'), ('with', 'O'), ('a', 'O'), ('4', 'O'), ('day', 'O'), ('history', 'O'), ('of', 'O'), ('facial', 'B-ety'), ('puffiness', 'I-ety'), (',', 'O'), ('peripheral', 'O'), ('pitting', 'B-ety'), ('oedema', 'I-ety'), (',', 'O'), ('abdominal', 'O'), ('pains', 'O'), (',', 'O'), ('and', 'O'), ('reduced', 'O'), ('urine', 'O'), ('output', 'O'), ('.', 'O')]


## Train-dev-test split
Notice that the original dataset has no dedicated validation split.

Randomly split the test set in 50% validation and 50% test

In [ ]:
print(len(train_set), len(dev_set), len(test_set))

669 425 426


## Sequence labels

Because BERT has a subword tokenizer, we need to map the label sequence to the subwords.

Beware to change the label from B to I for subwords that are in the middle of a word.

For example `'pit', '##ting'` should be tagged `'B-ety', 'I-ety'`, not `'B-ety', 'B-ety'`

Also make sure to add BERT special tokens for beginning of sequence and end of sequence. Give them `-100` as label so they are not taken into account when computing the loss.


In [ ]:
tokenizer.tokenize(" ".join(item[0]))

['A',
 '14',
 '-',
 'year',
 'old',
 'boy',
 'with',
 'no',
 'significant',
 'past',
 'medical',
 'history',
 'presented',
 'to',
 'a',
 'small',
 'district',
 'hospital',
 'in',
 'southern',
 'Sie',
 '##rr',
 '##a',
 'Leon',
 '##e',
 'with',
 'a',
 '4',
 'day',
 'history',
 'of',
 'facial',
 'pu',
 '##ffi',
 '##ness',
 ',',
 'peripheral',
 'pit',
 '##ting',
 'o',
 '##edema',
 ',',
 'abdominal',
 'pain',
 '##s',
 ',',
 'and',
 'reduced',
 'urine',
 'output',
 '.']

In [ ]:
tokenizer.cls_token

'[CLS]'

In [ ]:
tokenizer.sep_token

'[SEP]'

In [ ]:
# expected output

[('[CLS]', None), ('A', 'O'), ('14', 'O'), ('-', 'O'), ('year', 'O'), ('old', 'O'), ('boy', 'O'), ('with', 'O'), ('no', 'O'), ('significant', 'O'), ('past', 'O'), ('medical', 'O'), ('history', 'O'), ('presented', 'O'), ('to', 'O'), ('a', 'O'), ('small', 'O'), ('district', 'O'), ('hospital', 'O'), ('in', 'O'), ('southern', 'O'), ('Sie', 'O'), ('##rr', 'O'), ('##a', 'O'), ('Leon', 'O'), ('##e', 'O'), ('with', 'O'), ('a', 'O'), ('4', 'O'), ('day', 'O'), ('history', 'O'), ('of', 'O'), ('facial', 'B-ety'), ('pu', 'I-ety'), ('##ffi', 'I-ety'), ('##ness', 'I-ety'), (',', 'O'), ('peripheral', 'O'), ('pit', 'B-ety'), ('##ting', 'I-ety'), ('o', 'I-ety'), ('##edema', 'I-ety'), (',', 'O'), ('abdominal', 'O'), ('pain', 'O'), ('##s', 'O'), (',', 'O'), ('and', 'O'), ('reduced', 'O'), ('urine', 'O'), ('output', 'O'), ('.', 'O'), ('[SEP]', None)]


## Batching

In [ ]:
batch_size = 4
train_loader = torch.utils.data.DataLoader(train_set, batch_size=batch_size, shuffle=True, collate_fn=lambda x: x)


In [ ]:
batch = next(iter(train_loader))

Tokenize and get labels for all examples in the batch. Get the identifiers of tokens using `tokenizer.convert_tokens_to_ids`

Save the length (number of tokens) of each example in a separate list

You should end up with a list of list like so

In [ ]:
tokens_batch

[[101, 22690, 106, 163, 3055, 28224, 171, 211, 102],
 [101,
  5879,
  578,
  813,
  136,
  3174,
  864,
  689,
  806,
  603,
  6786,
  23642,
  855,
  111,
  803,
  939,
  2015,
  211,
  102],
 [101,
  186,
  22950,
  253,
  105,
  2579,
  953,
  2477,
  125,
  198,
  28351,
  30106,
  1003,
  6118,
  8970,
  2212,
  7108,
  15307,
  7290,
  211,
  102],
 [101,
  186,
  1476,
  11732,
  1337,
  125,
  105,
  5632,
  18459,
  430,
  596,
  907,
  537,
  11708,
  430,
  15184,
  968,
  430,
  251,
  9551,
  11012,
  7026,
  211,
  102]]

In [ ]:
labels_batch

[[-100, 0, 0, 0, 0, 0, 0, 0, -100],
 [-100, 0, 0, 0, 0, 0, 0, 1, 2, 2, 2, 0, 0, 0, 0, 0, 0, 0, -100],
 [-100, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -100],
 [-100,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
  0,
  1,
  2,
  0,
  0,
  1,
  2,
  2,
  0,
  -100]]

Notice that each example in the batch has a different size so we will pad them. The padded tokens will be masked in the attention mechanism (already implemented in BERT, you simply need to pass `attention_mask`) and will be given the label `-100` so they are not taken into account when computing the loss.


In [ ]:
lengths

[9, 19, 21, 24]

Here we simply add padding so that all examples can fit in the same Tensor

In [ ]:
input_ids = torch.zeros(len(batch), max(lengths), dtype=int, device="cuda")
attention_mask = torch.zeros(len(batch), max(lengths), dtype=int, device="cuda")
labels = torch.full((len(batch), max(lengths)), -100, dtype=int, device="cuda")

for i, (token, label) in enumerate(zip(tokens_batch, labels_batch)):
    input_ids[i, :len(token)] = torch.tensor(token)
    attention_mask[i, :len(token)] = 1
    labels[i, :len(label)] = torch.tensor(label)

Voilà!

In [ ]:
input_ids.shape, attention_mask.shape, labels.shape

(torch.Size([4, 24]), torch.Size([4, 24]), torch.Size([4, 24]))

In [ ]:
input_ids

tensor([[  101, 22690,   106,   163,  3055, 28224,   171,   211,   102,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0],
        [  101,  5879,   578,   813,   136,  3174,   864,   689,   806,   603,
          6786, 23642,   855,   111,   803,   939,  2015,   211,   102,     0,
             0,     0,     0,     0],
        [  101,   186, 22950,   253,   105,  2579,   953,  2477,   125,   198,
         28351, 30106,  1003,  6118,  8970,  2212,  7108, 15307,  7290,   211,
           102,     0,     0,     0],
        [  101,   186,  1476, 11732,  1337,   125,   105,  5632, 18459,   430,
           596,   907,   537, 11708,   430, 15184,   968,   430,   251,  9551,
         11012,  7026,   211,   102]], device='cuda:0')

In [ ]:
labels

tensor([[-100,    0,    0,    0,    0,    0,    0,    0, -100, -100, -100, -100,
         -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100],
        [-100,    0,    0,    0,    0,    0,    0,    1,    2,    2,    2,    0,
            0,    0,    0,    0,    0,    0, -100, -100, -100, -100, -100, -100],
        [-100,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0, -100, -100, -100, -100],
        [-100,    0,    0,    0,    0,    0,    0,    0,    1,    0,    0,    0,
            0,    1,    0,    1,    2,    0,    0,    1,    2,    2,    0, -100]],
       device='cuda:0')

In [ ]:
attention_mask

tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]],
       device='cuda:0')

# Model
## Encoding with BERT

BERT provides contextualized embeddings for all tokens in the batch

In [ ]:
output = encoder(input_ids=input_ids, attention_mask=attention_mask)

In [ ]:
# (batch size, sequence length, embedding dimension)
output.last_hidden_state.shape

torch.Size([4, 39, 768])

## Sequence Tagging
For Named Entity Recognition, we use a simple tagging model: we simply add a linear layer on top of the encoder for multi-class classification:
- What is the input dimension of the classifier?
- How many classes are there?

Note: this Linear classifier is really simple and **assigns a label to each token independently**. A better way to do Named Entity Recognition that we did not cover in class is to use Conditional Random Field (CRF) for **structured prediction** (basically, assign the label of a token based on the other labels already assigned).

Note 2: HuggingFace's `transformers` provides built-in sequence tagging models like https://huggingface.co/docs/transformers/model_doc/bert#transformers.BertForTokenClassification which you are not allowed to use until the end of this class

In [ ]:
class Tagger(nn.Module):
    def __init__(self, encoder):
        super().__init__()
        self.encoder = encoder
        raise NotImplementedError()
        self.classifier = TODO

    def forward(self, input_ids, attention_mask):
        embeddings = self.encoder(input_ids=input_ids, attention_mask=attention_mask).last_hidden_state
        raise NotImplementedError()

In [ ]:
tagger = Tagger(encoder).cuda()

In [ ]:
logits = tagger(input_ids, attention_mask)

In [ ]:
# (batch size, sequence length, number of classes)
logits.shape

torch.Size([4, 24, 3])

## Loss function

Use Cross-entropy to train the classifier. First compute the loss on a single batch.

In [ ]:
# value may vary but you might get a loss like this one

tensor(1.0238, device='cuda:0', grad_fn=<NllLossBackward0>)

# Training

In [ ]:
%load_ext tensorboard

In [ ]:
import torch
from torch.utils.tensorboard import SummaryWriter
writer = SummaryWriter("logs/pw4")

Run tensorboard before training. Refresh during training.

In [ ]:
%tensorboard --logdir logs/pw4

In [ ]:
tagger = Tagger(encoder).cuda()
# gradient checkpoint: lower memory footprint but you need to compute forward passes twice
tagger.encoder.gradient_checkpointing_enable()

loss_fct = nn.CrossEntropyLoss(ignore_index=-100)



optimizer = torch.optim.AdamW(tagger.parameters(), lr=0.0001)

batch_size = 8
train_loader = torch.utils.data.DataLoader(train_set, batch_size=batch_size, shuffle=True, collate_fn=lambda x: x)
validation_loader = torch.utils.data.DataLoader(dev_set, batch_size=batch_size, shuffle=False, collate_fn=lambda x: x)

steps = 0
for epoch in range(5):
    for batch in train_loader:
        raise NotImplementedError("Format batch and compute loss as you did above")
        loss = TODO

        writer.add_scalar("Loss/train", loss.item(), steps)
        steps += 1
        loss.backward()
        # gradient clipping to avoid exploding gradients
        nn.utils.clip_grad_norm_(tagger.parameters(), 1.0)
        optimizer.step()

    # validation
    with torch.no_grad():
        tagger.eval()
        valid_loss = 0
        valid_batches = 0
        for batch in validation_loader:
            raise NotImplementedError("Format batch and compute loss as you did above")
            loss = TODO
            valid_loss += loss.item()
            valid_batches += 1
        tagger.train()
        writer.add_scalar("Loss/validation", valid_loss/valid_batches, epoch)

    # saving checkpoint
    torch.save(tagger.state_dict(), f"tagger_{epoch}.pt")

# Testing

Notice that the model quickly overfits (after 1 epoch?) given the small size of the training steps. Load the best checkpoint according to the validation loss (epoch 1 in my case)

In [ ]:
tagger.load_state_dict(torch.load("tagger_0.pt"))
tagger.eval()

/tmp/ipykernel_9055/4253622363.py:1: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  tagger.load_state_dict(torch.load("tagger_0.pt"))


Tagger(
  (encoder): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(31116, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-11): 12 x BertLayer(
          (attention): BertAttention(
            (self): BertSdpaSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_aff

In [ ]:
LABELS

{'O': 0, 'B-ety': 1, 'I-ety': 2}

Now run your model on your test set. This time we're not only interested in a low loss but in the actual predictions of the models:
- How to get the class predicted by your model on each token?

Once you get classes, compute:
- precision (of classifying 'B-ety' and 'I-ety')
- recall (of classifying 'B-ety' and 'I-ety')
- F1-score (of classifying 'B-ety' and 'I-ety')

Bonus: what is the issue with computing scores on tokens tokenized using BERT tokenizer?

In [ ]:
# example prediction of my model

[('[CLS]', 'O'),
 ('Given', 'O'),
 ('the', 'O'),
 ('positive', 'O'),
 ('screen', 'O'),
 ('for', 'O'),
 ('ce', 'B-ety'),
 ('##li', 'I-ety'),
 ('##ac', 'I-ety'),
 ('disease', 'I-ety'),
 ('(', 'O'),
 ('positive', 'O'),
 ('anti', 'O'),
 ('-', 'O'),
 ('tissue', 'O'),
 ('trans', 'O'),
 ('##glut', 'O'),
 ('##aminase', 'O'),
 ('antibodies', 'O'),
 ('and', 'O'),
 ('results', 'O'),
 ('of', 'O'),
 ('duoden', 'O'),
 ('##al', 'O'),
 ('biopsy', 'O'),
 (')', 'O'),
 (',', 'O'),
 ('dietary', 'O'),
 ('intervention', 'O'),
 ('was', 'O'),
 ('immediately', 'O'),
 ('comm', 'O'),
 ('##enced', 'O'),
 ('.', 'O'),
 ('[SEP]', 'O')]

In [ ]:
test_loader = torch.utils.data.DataLoader(test_set, batch_size=batch_size, shuffle=False, collate_fn=lambda x: x)
for batch in test_loader:
    raise NotImplementedError()

The result is pretty bad but consider we use a very small training set

In [ ]:
print(f"{precision=:.1%}, {recall=:.1%}, {f_score=:.1%}")

precision=61.3%, recall=29.3%, f_score=39.7%


# Bonus

## Evaluation

- Resplit the data by keeping the 426 longest texts as out-of-distribution test set. The remaining can be randomly split in train-validation.
- Compute F1 score using detokenized outputs

## Parameter-efficient fine-tuning

Compare full fine-tuning like above with LoRA

## More encoders

Evaluate other models, e.g.:
- another encoder, e.g. BERT or RoBERTa that were trained on general domain texts or https://huggingface.co/medicalai/ClinicalBERT that was trained on clinical texts
- a decoder-only like GPT-2
- an encoder-decoder like BART

## Better classifiers

use Conditional Random Field (CRF) instead of a linear classifier